linalg/ode.hpp

```diff
@@ -12,13 +12,91 @@
 #ifndef MFEM_ODE
 #define MFEM_ODE
 
+#include "../general/communication.hpp"
 #include "../config/config.hpp"
 #include "operator.hpp"
-#include "../general/communication.hpp"
+#include <vector>
+#include <memory>
 
 namespace mfem
 {
 
+/// An interface for storing state of previous timesteps
+class ODEStateData
+{
+public:
+   /// Get the maximum number of stored stages
+   virtual int MaxSize() const = 0;
+
+   /// Get the current number of stored stages
+   virtual int Size() const = 0;
+
+   /// Get the ith state vector
+   virtual void Get(int i, Vector &state) const = 0;
+
+   /// Get the ith state vector
+   virtual const Vector &Get(int i) const = 0;
+
+   /// Set the ith state vector
+   virtual void Set(int i, Vector &state) = 0;
```

Comment:
To avoid unnecessary copies, isn't it better to provide non-const `virtual Vector &Get(int) = 0` method? It will replace this method which I suspect copies the given `state` to an internal `Vector`.

Reply:
I am inclined to think this function will break the generality aimed for by @cjvogl.
What is your opinion on this @cjvogl?

Reply:
The function signature here was motivated by the former `ODESolver::SetStateVector`. Aside from that, I see the point in wanting to avoid copies in either `ODESolver::SetStateVector` or in this new interface. 

Given the existing function `const Vector &Get(int)` effectively requires derived classes to implement storage of states as `Vector` objects or something that can be represented as `Vector` objects, I don't see adding a non-const version of `Get` to the interface as imposing any additional requirements on derived classes. Furthermore, I suppose a derived class can always overload the non-const `Get` with something that throws an error if that derived class wants to restrict state data changes through `Set`. (e.g., if a flag or something is to be set when the state vectors are changed).

Just thinking out loud here, if someone down the road wants to implement a derived class that doesn't store states as `Vector` objects, this interface can be generalized to provide something like 
```c++
template<typename T> class ODEStateData
{ ...
   virtual T& Get(i);
   const virtual T& Get(i) const;
}